src/lib/codec/hex/hex.cpp

```diff
@@ -53,27 +60,36 @@ std::string hex_encode(const uint8_t input[], size_t input_length, bool uppercas
 namespace {
 
 uint8_t hex_char_to_bin(char input) {
-   const uint8_t c = static_cast<uint8_t>(input);
+   // Assumes each byte is either 0x00 or 0x80
+   auto index_of_first_set_byte = [](uint64_t v) {
+      return ((((v - 1) & 0x0101010101010101) * 0x0101010101010101) >> 56) - 1;
+   };
+
+   auto swar_in_range = [](uint64_t v, uint64_t lower, uint64_t upper) -> uint64_t {
+      // This assumes that the high bits of each byte of upper/lower are unset
+      constexpr uint64_t hi = 0x8080808080808080;
+      constexpr uint64_t lo7 = 0x7F7F7F7F7F7F7F7F;
 
-   const auto is_alpha_upper = CT::Mask<uint8_t>::is_within_range(c, uint8_t('A'), uint8_t('F'));
-   const auto is_alpha_lower = CT::Mask<uint8_t>::is_within_range(c, uint8_t('a'), uint8_t('f'));
-   const auto is_decimal = CT::Mask<uint8_t>::is_within_range(c, uint8_t('0'), uint8_t('9'));
+      const uint64_t sub = ((v | hi) - (lower & lo7)) ^ ((v ^ (~lower)) & hi);
+      const uint64_t a_lo = sub & lo7;
+      const uint64_t a_hi = sub & hi;
+      return (lo7 - a_lo + upper) & hi & ~a_hi;
+   };
```

Comment:
This (and perhaps other helpers) might be a good candidate for reusable SWAR utilities. On a first glance this looks very similar to the implementation in base64. Or am I missing the fine details?

Reply:
I hesitated on these since the way these are formulated is not totally general (in that they cannot handle inputs where the high bit is set). This can be done but adds several more operations which I’d like to avoid given that the whole point is to make encoding/decoding faster than the current CT::Mask code. But they are otherwise basically identical (at most differing based on the size of the type in use) so I moved these to `int_utils.h` with doc comments indicating the current limitations.